k2/csrc/thread_pool.h

```diff
@@ -0,0 +1,108 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+#ifndef K2_CSRC_THREAD_POOL_H_
+#define K2_CSRC_THREAD_POOL_H_
+
+#include <condition_variable>  // NOLINT
+#include <functional>
+#include <mutex>  // NOLINT
+#include <queue>
+#include <thread>  // NOLINT
+#include <vector>
+
+#include "k2/csrc/log.h"
+
+namespace k2 {
+
+class ThreadPool {
+ public:
+  using TaskFunc = std::function<void()>;
+
+  /* Create a pool with given number of threads.
+   *
+   * @param [in] num_threads  The number of threads in the pool. If it is <= 0,
+   *                          the number of threads in the pool is set to
+   *                          `std::thread::hardware_concurrency()`.
+   */
+  explicit ThreadPool(int32_t num_threads);
+
+  /* It ask all threads in the pool to exit after processing what they
+   * are currently executing. If there are still tasks in the queue, they
+   * will not be executed.
+   */
+  ~ThreadPool();
```

Comment:
I suggest we make assignment operator and copy constructor  deleted, both for lvalue and rvalue (move), since it's dangerous to copy/assgin a thread pool object.

Reply:
> since it's dangerous to copy/assgin a thread pool object.

I don't think you are able to copy or assign it.

Reply:
I added some comments explaining why it is not copyable or assignable.